tests/functional/test_build_constraints.py

```diff
@@ -0,0 +1,176 @@
+    project_dir: Path,
+    constraints_file: Path,
+) -> TestPipResult:
+    """Run pip install with build constraints but without the feature flag."""
+    return script.pip(
+        "install",
+        "--build-constraint",
+        str(constraints_file),
+        str(project_dir),
+    )
+
+
+def test_build_constraints_basic_functionality_simple(
+    script: PipTestEnvironment, tmpdir: Path
+) -> None:
+    """Test that build constraints options are accepted and processed."""
+    project_dir = _create_simple_test_package(
+        script=script, name="test_build_constraints"
+    )
+    constraints_file = _create_constraints_file(
+        script=script, filename="constraints.txt", content="setuptools>=40.0.0\n"
+    )
+    result = _run_pip_install_with_build_constraints(
+        script=script, project_dir=project_dir, build_constraints_file=constraints_file
+    )
+    _assert_successful_installation(
+        result=result, package_name="test_build_constraints"
+    )
+
+
+@pytest.mark.network
+def test_build_constraints_vs_regular_constraints_simple(
+    script: PipTestEnvironment, tmpdir: Path
+) -> None:
+    """Test that build constraints and regular constraints work independently."""
+    project_dir = create_test_package_with_setup(
+        script,
+        name="test_isolation",
+        version="1.0",
+        py_modules=["test_isolation"],
+        install_requires=["six"],
+    )
+    build_constraints_file = _create_constraints_file(
+        script=script, filename="build_constraints.txt", content="setuptools>=40.0.0\n"
+    )
+    regular_constraints_file = _create_constraints_file(
+        script=script, filename="constraints.txt", content="six>=1.10.0\n"
+    )
+    result = script.pip(
+        "install",
+        "--no-cache-dir",
+        "--build-constraint",
+        build_constraints_file,
+        "--constraint",
+        regular_constraints_file,
+        "--use-feature",
+        "build-constraint",
+        str(project_dir),
+        expect_error=False,
+    )
+    assert "Successfully installed" in result.stdout
+    assert "test_isolation" in result.stdout
+
+
+@pytest.mark.network
+def test_build_constraints_environment_isolation_simple(
+    script: PipTestEnvironment, tmpdir: Path
+) -> None:
+    """Test that build constraints work correctly in isolated build environments."""
+    project_dir = _create_simple_test_package(script=script, name="test_env_isolation")
+    constraints_file = _create_constraints_file(
+        script=script, filename="build_constraints.txt", content="setuptools>=40.0.0\n"
+    )
+    result = _run_pip_install_with_build_constraints(
+        script=script,
+        project_dir=project_dir,
+        build_constraints_file=constraints_file,
+        extra_args=["--isolated"],
+    )
+    _assert_successful_installation(result=result, package_name="test_env_isolation")
+
+
+def test_build_constraints_file_not_found(
+    script: PipTestEnvironment, tmpdir: Path
+) -> None:
```

Comment:
What is the reason why `--build-constraint` will fail silently if the file isn't found? This deviates from `--constraint` which will loudly fail if the file can't be read.